### PR TITLE
Added more tests

### PR DESCRIPTION
**Pull request by abhishekchaturvedi**
Changed int to int64 and added tests for string as keys for the same operations.

### CODE REVIEW

File: bplustree.go [hunk 1]
```diff
@@ -61,10 +61,6 @@ func (lck *defaultLock) Unlock(readSet []Key, writeSet []Key) {
 // 'Load' is a function that should yield all key,value pairs iteratively for
 // so that the BplusTree could be loaded with the key/values in the db when the
 // BplusTree is instantiated.
-// The typical usage of the loader will look like:
```
Comment (ajayguls) [inline bplustree.go:64]: line 32: readseys -> readSet
line 44: if len(readSet) > 0 or len(writeSet) > 0
same for line 51
add a new line between lock related functions
comment from line 61 to 63 also needs to be changed.

Load() has to take key as a parameter. Otherwise how will the code load the node specific to a key.
Also the key of a node that needs to be stored is different from the actual keys in the tree. The key space for the tree nodes is kept in the actual database like arango. We can use a fixed type for that. it doesn't have to be flexible.

For MemMgr:
I think the interface should be Insert(key, value), Lookup(key) and Delete(key)
if lookup fails, then the code will read from DBMgr and then insert to MemMap

File: bplustree_test.go [hunk 2]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Got nil result")
+		t.FailNow()
+	}
+
+	// case 1.a Exact search. Provide an evaluator and maxElems which
+	// should be ignored.
+	testingKey1 := makeTestingKeyStr(100)
```
Comment (ajayguls) [inline bplustree_test.go:615]: pick a random element in the range instead of a fixed one like 100.

File: bplustree_test.go [hunk 3]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	result := k1.Compare(k2)
+	if result < 10 && result > -10 {
+		return true
+	}
+
+	return false
+}
+func TestSearchStr1(t *testing.T) {
+	// case 1. Exact search.
+	testingKey := makeTestingKeyStr(100)
```
Comment (ajayguls) [inline bplustree_test.go:601]: pick a random element instead of 100

File: bplustree_test.go [hunk 4]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	if result != nil {
+		t.Errorf("Expected to not get any result but got: %v", result)
+		t.FailNow()
+	}
+
+}
+
+func keyEvaluatorStr(k1 Key, k2 Key) bool {
+	result := k1.Compare(k2)
+	if result < 10 && result > -10 {
```
Comment (ajayguls) [inline bplustree_test.go:593]: what is the significance of 10 and -10? is this to restrict the evaluations that would yield true? so if you ask for 30 results, you will get maximum of 20?

File: bplustree_test.go [hunk 5]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+		t.FailNow()
+	}
+	if err != nil {
+		t.Errorf("Expected to find the key: %v (err: %v)", ss.searchKey, err)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Got nil result")
+		t.FailNow()
+	}
```
Comment (ajayguls) [inline bplustree_test.go:630]: check that the result actually matches with what is expected.
same for int64 case also.

File: bplustree_test.go [hunk 6]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+}
+
+func TestSearchStr2(t *testing.T) {
+	// case 2. maxElems greater than limit.
```
Comment (ajayguls) [inline bplustree_test.go:634]: what is the limit here? where is that defined?

File: bplustree_test.go [hunk 7]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Exact,
+		maxElems: 100, evaluator: keyEvaluatorStr}
+
+	result, err := newtree.Search(ss)
+	if err == nil {
+		t.Errorf("Expected to not find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
```
Comment (ajayguls) [inline bplustree_test.go:643]: check the error type to be what is expected.

File: bplustree_test.go [hunk 8]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	// case 3. maxElems = 10, in Right direction. nil evaluator
+	testingKey := makeTestingKeyStr(100)
+	ss := SearchSpecifier{searchKey: testingKey, direction: Right,
+		maxElems: 10}
+
+	result, err := newtree.Search(ss)
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
```
Comment (ajayguls) [inline bplustree_test.go:660]: all such cases can be replaced by a single line, you can use a package for asserts.

File: bplustree_test.go [hunk 9]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
```
Comment (ajayguls) [inline bplustree_test.go:664]: same for these. I will send the name of package, you can use that.

File: bplustree_test.go [hunk 10]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+
+	if len(result) < ss.maxElems || len(result) > ss.maxElems+1 {
```
Comment (ajayguls) [inline bplustree_test.go:666]: why +1 here? why not check that len(results) != ss.maxElems

File: bplustree_test.go [hunk 11]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestSearchStr4(t *testing.T) {
+	// case 4. maxElems = 10, in Left direction.
+	testingKey := makeTestingKeyStr(100)
```
Comment (ajayguls) [inline bplustree_test.go:687]: use a different / random key instead of 100

File: bplustree_test.go [hunk 12]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) < ss.maxElems || len(result) > ss.maxElems+1 {
```
Comment (ajayguls) [inline bplustree_test.go:701]: why not check for equality instead of this dual check?

File: bplustree_test.go [hunk 13]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	// Note that our values are +10 of the key.
+	startKey := makeTestingKeyStr(strHash(testingKey) - 10)
+	expectedVal := int(strHash(startKey) + 10)
+	for i := 0; i < len(result); i++ {
+		//fmt.Println(reflect.TypeOf(result[i]))
+		actualVal := result[i].(*TestingElemStr).val
+		if actualVal != expectedVal {
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
```
Comment (ajayguls) [inline bplustree_test.go:750]: all these can be single line also with the assert package

File: bplustree_test.go [hunk 14]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) < ss.maxElems || len(result) > ss.maxElems+1 {
```
Comment (ajayguls) [inline bplustree_test.go:784]: this will pass for 2 values of len(result), don't we expect a single value?

File: bplustree_test.go [hunk 15]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Expected to find the key: %v", ss.searchKey)
+		t.FailNow()
+	}
+	if result == nil {
+		t.Errorf("Expected to get result but nil: %v", result)
+		t.FailNow()
+	}
+
+	if len(result) != 5 {
```
Comment (ajayguls) [inline bplustree_test.go:832]: how do you expect the len() to be 5?

File: bplustree_test.go [hunk 16]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+			t.Errorf("At index %d: expected: %d, got %v", i, expectedVal, actualVal)
+			t.FailNow()
+		}
+		// and increment by 1.
+		expectedVal++
+	}
+}
+
+func TestRemoveStr1(t *testing.T) {
+	fmt.Println("Starting remove in order tests")
```
Comment (ajayguls) [inline bplustree_test.go:853]: are you expecting the test functions to run in order?  may not be a safe assumption if someone adds code in the file or moves things around.

File: bplustree_test.go [hunk 17]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	fmt.Println("Starting remove in reverse order tests")
+	for i := numElems - 1; i >= 0; i-- {
+		key := makeTestingKeyStr(i)
+		fmt.Printf("removing key: %v", key)
+		err := newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+	}
```
Comment (ajayguls) [inline bplustree_test.go:876]: do lookup for some elements to make sure nothing is left. also would be useful to do some lookups of existing and deleted keys during the remove to make sure lookups continue to work

File: bplustree_test.go [hunk 18]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+
+		k3 := numElems*2/3 + i
+		key = makeTestingKeyStr(k3)
+		fmt.Printf("removing key: %v", key)
+		err = newtree.Remove(key)
+		if err != nil {
+			t.Errorf("Failed to remove key: %v got %v", key, err)
+			t.FailNow()
+		}
+		removed[k3] = true
```
Comment (ajayguls) [inline bplustree_test.go:913]: add some lookups here to make sure the tree is working as expected.

File: bplustree_test.go [hunk 19]
```diff
@@ -464,3 +515,437 @@ func TestRemove4(t *testing.T) {
...
+	}
+
+	wg.Wait()
+	for i := 0; i < numElems; i++ {
+		if errors[i] != nil {
+			t.Errorf("Failed to insert: %d: %v", i, errors[i])
+			t.FailNow()
+		}
+	}
+	t.Logf("\nFinal tree:\n")
```
Comment (ajayguls) [inline bplustree_test.go:949]: you are mostly removing all elements. how about adding cases to make sure rebalancing happens properly on removals that cause numelements to be maxsize/2?
I feel that the edge cases are not being test here.

**Review by ajayguls — COMMENTED** [review]
Some comments below.

File: bplustree.go (hunk 1 above)
Comment (abhishekchaturvedi) [reply bplustree.go:64]: Sure.. that was just a last minute change to remove some comment.. The interface needs a bit of rework.. look for changes regarding these in changes when we start bringing in memMgr implementation.. I also think that we need a bit more discussion regarding memMgr.. I will call u today.

File: bplustree_test.go (hunk 2 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:615]: sure.

File: bplustree_test.go (hunk 3 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:601]: sure

File: bplustree_test.go (hunk 4 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:593]: yeah.. if evaluator stops yielding then maxElem doesn't matter.. otherwise maxElems will limit the # of items fetched. User can define the evaluator to their choice.. in our usecase I expect we could have the evaluator which says, give me keys until # of bits difference is less than certain value.. or me 50 closeby keys .. whichever is less.

File: bplustree_test.go (hunk 5 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:630]: sure.

File: bplustree_test.go (hunk 16 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:853]: Unless I write a giant function which does everything.. it has to be this way.. can't expect to remove things if they aren't 'insert'ed. And if tests are added in order that it breaks something then the test will break right away and whoever adds new test would know. And I just don't want to have 1 single function with all functionality. During multiple remove tests through.. I am reinserting the keys,, so no ordering is needed.. I could make each testcase be self sufficient (unlike only some of them which are), but then inserting 100K entries for each test case becomes an overkill..

File: bplustree_test.go (hunk 17 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:876]: sure.

File: bplustree_test.go (hunk 18 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:913]: added

File: bplustree_test.go (hunk 19 above)
Comment (abhishekchaturvedi) [reply bplustree_test.go:949]: They are guaranteed to happen.. there's no way to validate that redistribute/merge happen programatically.. but going in the deterministic way, like in order or in reverse order, we are guaranteed to have both these occur upon deletion of of consecutive 'degree' # of elements. in order and reverse order also guarantee that we are covering all edge cases of prev/next etc. being null in certain conditions. And also I have things being removed from middle also every now and then..
